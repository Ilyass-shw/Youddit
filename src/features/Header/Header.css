.header {
	display: inline-flex;
	align-items: center;
	margin-left: 4.5rem;
	height: 3.8rem;
	border-block-end: 1px solid #404040;
}
.reddit-logo {
	flex: 1;
	display: flex;
	justify-content: flex-start;
	min-width: 8rem;
}
.reddit-logo h3 {
	margin-left: 0.6rem;
}

.go-back-button {
	padding: 0.5rem;
	border-radius: 50%;
}
.inputForm,
.header-user-features {
	flex: 1;
}
.input {
	flex: 3;
	min-width: 0;
}
.header-user-features {
	flex: 1;
	display: inline-flex;
	justify-content: space-evenly;
	align-items: center;
	margin-left: 20.19px;
	margin-right: 20.19px;
}
.icon-container {
	position: relative;
}
.user-icon {
	padding: 0.5rem;
}
.icon-label {
	padding: 0.4rem;
	color: hsl(0, 0%, 100%);
	background-color: rgba(90, 90, 90, 0.788);
	border-radius: 2px;
	position: absolute;
	top: 30px;
	right: -4px;
	opacity: 0;
	transition: opacity 0.28s;
	font-size: 0.75rem;
	z-index: 3;
}

.user-icon:hover + p {
	opacity: 1;
}
