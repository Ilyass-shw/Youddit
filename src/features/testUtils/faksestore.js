import { configureStore } from "@reduxjs/toolkit";
import { createSlice } from "@reduxjs/toolkit";

const initialState = {
	firstFilter: "relevance",
	lastFilter: "comments",
	filter: "relevance",
};
const fakeSlice = createSlice({
	name: "posts",
	initialState,
	reducers: {},
});

export default configureStore({
	reducer: {
		posts: fakeSlice.reducer,
	},
});

export const makeCustomTestStore = (name, initialState, reducers = {}) => {
	const fakeSlice = createSlice({
		name,
		initialState,
		reducers,
	});

	return configureStore({
		reducer: {
			posts: fakeSlice.reducer,
		},
	});
};
