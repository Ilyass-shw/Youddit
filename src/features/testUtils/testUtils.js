import React from "react";
import { render as rtlRender } from "@testing-library/react";
import { Provider } from "react-redux";
import store, { makeCustomTestStore } from "./faksestore";

const TestProvider = ({ store, children }) => <Provider store={store}>{children}</Provider>;

export function render(ui, { store, ...otherOpts }) {
	return rtlRender(<TestProvider store={store}>{ui}</TestProvider>, otherOpts);
}

export function makeTestStore(name, initialState, reducers) {
	if (name && initialState) {
		const Store = makeCustomTestStore(name, initialState, reducers);
		const origDispatch = Store.dispatch;
		Store.dispatch = jest.fn(origDispatch);
		return Store;
	} else {
		const origDispatch = store.dispatch;
		store.dispatch = jest.fn(origDispatch);
		return store;
	}
}

// re-export everything
export * from "@testing-library/react";
