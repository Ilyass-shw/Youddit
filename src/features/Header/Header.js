import React, { useState } from "react";
import "./Header.css";
import { InlineIcon } from "@iconify/react";
import redditFill from "@iconify-icons/akar-icons/reddit-fill";
import bxsVideoPlus from "@iconify-icons/bx/bxs-video-plus";
import appMenu from "@iconify-icons/fe/app-menu";
import bellIcon from "@iconify-icons/mdi/bell";
import { InputForm } from "../InputForm/InputForm.js";
import search16Filled from "@iconify-icons/fluent/search-16-filled";
import arrowLeft from "@iconify-icons/akar-icons/arrow-left";
import useWindowDimensions from "../Helper/UseWindowDimensions";
const Header = () => {

	const {width} = useWindowDimensions()
	let isNotDesktop =width< 460;

	const [isSearching, setIsSearching] = useState(false);

	const handleOnSearchClick = () => {
		isNotDesktop = false;
		setIsSearching(true);
	};
	const handleOnBackArrowClick = () => {
		setIsSearching(false);
	};

	return (
		<div className="header">
			{!isSearching && (
				<div className="reddit-logo" data-testid="reddit-logo">
					<h3>
						<InlineIcon icon={redditFill} color={"red"} width="1.7rem" height="1.7rem" /> Tubeddit
					</h3>
				</div>
			)}

			{isSearching && (
				<div className="go-back-button" data-testid="go-back-button" onClick={handleOnBackArrowClick}>
					<InlineIcon icon={arrowLeft} />
				</div>
			)}

			{(!isNotDesktop || isSearching) && <InputForm className="inputForm" />}

			{!isSearching && (
				<div className="header-user-features">
					<div className="icon-container">
						{isNotDesktop && (
							<button data-testid="mobile-search-button" className="mobile-search-button" onClick={handleOnSearchClick}>
								<InlineIcon icon={search16Filled} width="1.5rem" height="1.5rem" color={"white"} />
							</button>
						)}
					</div>

					<div className="icon-container">
						<a className="user-icon" href="www.youtube.com/watch?v=3oUu7N65s6I&t=1860s">
							<InlineIcon icon={bxsVideoPlus} color={"white"} width="1.5rem" height="1.5rem" />
						</a>
						<p className="icon-label">Create</p>
					</div>

					<div className="icon-container">
						<a className="user-icon" href="www.youtube.com/watch?v=3oUu7N65s6I&t=1860s">
							<InlineIcon icon={appMenu} color={"white"} width="1.5rem" height="1.5rem" />
						</a>
						<p className="icon-label">Apps</p>
					</div>

					<div className="icon-container">
						<a className="user-icon" href="www.youtube.com/watch?v=3oUu7N65s6I&t=1860s">
							<InlineIcon icon={bellIcon} color={"white"} width="1.5rem" height="1.5rem" />
						</a>
						<p className="icon-label">Notifications</p>
					</div>

					<div className="icon-container">
						<a className="user-icon" href="www.youtube.com/watch?v=3oUu7N65s6I&t=1860s">
							<InlineIcon icon={redditFill} color={"white"} width="1.5rem" height="1.5rem" />
						</a>
					</div>
				</div>
			)}
		</div>
	);
};


export default Header;
