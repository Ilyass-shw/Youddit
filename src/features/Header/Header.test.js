import React from "react";
import Header from "./Header";
import { jest } from "@jest/globals";
import { render, makeTestStore, fireEvent, screen } from "../testUtils/testUtils";

jest.mock("@iconify/react", () => {
	return { InlineIcon: () => <p>InlineIcon mock</p> };
});
jest.mock("../InputForm/InputForm.js", () => {
	return { InputForm: () => <p>InputForm mock</p> };
});

jest.mock("@iconify-icons/akar-icons/reddit-fill", () => {
	return {
		default: () => {
			return "rock";
		},
	};
});

jest.mock("@iconify-icons/bx/bxs-video-plus", () => {
	return {
		default: () => {
			return "rock";
		},
	};
});

jest.mock("@iconify-icons/fe/app-menu", () => {
	return {
		default: () => {
			return "rock";
		},
	};
});

jest.mock("@iconify-icons/mdi/bell", () => {
	return {
		default: () => {
			return "rock";
		},
	};
});

jest.mock("@iconify-icons/akar-icons/arrow-left", () => {
	return {
		default: () => {
			return "rock";
		},
	};
});
jest.mock("@iconify-icons/fluent/search-16-filled", () => {
	return {
		default: () => {
			return "rock";
		},
	};
});

let store;

describe("Header", () => {
	beforeEach(() => {
		store = makeTestStore();
	});

	it("should match the snapshot", () => {
		const { container } = render(<Header />, { store });
		expect(container).toMatchSnapshot();
	});

	it("should render without crashing", () => {
		render(<Header />, { store });
	});

	it("should render InputForm when window-width > 460", () => {
		global.innerWidth = 461;
		const { getByText } = render(<Header />, { store });
		getByText("InputForm mock");
	});

	it("should not render InputForm on window-width < 460", () => {
		global.innerWidth = 459;
		render(<Header />, { store });
		const InputForm = screen.queryByText("InputForm mock");
		expect(InputForm).not.toBeInTheDocument(); // it doesn't exist
	});

	it("should render InputForm and go-back-button when clicking on search button on window-width < 460", () => {
		global.innerWidth = 459;
		const { getByTestId } = render(<Header />, { store });
		const MobileSearchButton = getByTestId("mobile-search-button");
		fireEvent.click(MobileSearchButton);

		const InputForm = screen.queryByText("InputForm mock");
		expect(InputForm).toBeInTheDocument();

		const GoBackButton = screen.queryByTestId("go-back-button");
		expect(GoBackButton).toBeInTheDocument();
	});

	it("should make InputForm disappear and rerender mobile-search-button and reddit-logo when clicking on go-back-button on window-width < 460", () => {
		global.innerWidth = 459;
		const { getByTestId } = render(<Header />, { store });
		let MobileSearchButton = getByTestId("mobile-search-button");
		fireEvent.click(MobileSearchButton);

		let GoBackButton = screen.queryByTestId("go-back-button");
		fireEvent.click(GoBackButton);

		MobileSearchButton = getByTestId("mobile-search-button");
		expect(MobileSearchButton).toBeInTheDocument();

		const redditLogo = getByTestId("reddit-logo");
		expect(redditLogo).toBeInTheDocument();
	});
});
