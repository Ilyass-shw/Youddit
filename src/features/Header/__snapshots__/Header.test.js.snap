// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should match the snapshot 1`] = `
<div>
  <div
    class="header"
  >
    <div
      class="reddit-logo"
      data-testid="reddit-logo"
    >
      <h3>
        <p>
          InlineIcon mock
        </p>
         React
      </h3>
    </div>
    <p>
      InputForm mock
    </p>
    <div
      class="header-user-features"
    >
      <div
        class="icon-container"
      />
      <div
        class="icon-container"
      >
        <a
          class="user-icon"
          href="/"
        >
          <p>
            InlineIcon mock
          </p>
        </a>
        <p
          class="icon-label"
        >
          Create
        </p>
      </div>
      <div
        class="icon-container"
      >
        <a
          class="user-icon"
          href="/"
        >
          <p>
            InlineIcon mock
          </p>
        </a>
        <p
          class="icon-label"
        >
          Apps
        </p>
      </div>
      <div
        class="icon-container"
      >
        <a
          class="user-icon"
          href="/"
        >
          <p>
            InlineIcon mock
          </p>
        </a>
        <p
          class="icon-label"
        >
          Notifications
        </p>
      </div>
      <div
        class="icon-container"
      >
        <a
          class="user-icon"
          href="/"
        >
          <p>
            InlineIcon mock
          </p>
        </a>
      </div>
    </div>
  </div>
</div>
`;
